Initial design doc for scheduler.

davidopp opened the pull request:
Fixes #6781

@dingh let me know what you think. Please feel free to send a PR to add stuff. This turned out to be a little more than an outline, but I tried to keep it simple.

Commit:
Initial design doc for scheduler.

mikedanese commented:
@HaiyangDING

mikedanese commented on docs/design/scheduler.md:
```diff
@@ -0,0 +1,46 @@
+The scheduler is extensible: the cluster administrator can choose which of the pre-defined
+scheduling policies to apply, and can add new ones. The built-in predicates and priorities are
+defined in [plugin/pkg/scheduler/predicates.go](../../plugin/pkg/scheduler/predicates.go) and
+[plugin/pkg/scheduler/priorities.go](../../plugin/pkg/scheduler/priorities.go), respectively.
```
This link and the link on the line above are broken.

davidopp replied:
Thanks, fixed

derekwaynecarr commented on docs/design/scheduler.md:
```diff
@@ -0,0 +1,46 @@
+
+The Kubernetes scheduler runs as a process alongside the other master
+components such as the API server. Its interface to the API server is to watch
+for Pods with an empty PodSpec.NodeName, and for each them to post a Binding
```
"Its interface to the API server is to watch for Pods with an empty PodSpec.NodeName, and for each pod, it posts a Binding indicating where the Pod should be scheduled."

davidopp replied:
Done

HaiyangDING commented on docs/design/scheduler.md:
```diff
@@ -0,0 +1,46 @@
+that rank the nodes that weren't filtered out by the predicate check. For example,
+if the PodSpec specified resource limits, then it favors the least-loaded
+machine (measured in terms of the sum of the limits of the Pods already on each node).
+Finally, the node with the highest priority is chosen
```
May we should say something about weight?

Each priority function is assigned with a weight (a positive number) representing how much this priority function would affected the final score when ranking each node.

davidopp replied:
I think we can have a later section where we talk about the algorithms in more detail. Based on @erictune 's comment I wasn't sure how much detail we should get into. So I wanted this PR to be high-level, talking only about things that are unlikely to change soon.

HaiyangDING replied:
I totally agree with you on this. The conception of weight itself is not hard to understand but introducing it here is likely to make this doc a little messy.

HaiyangDING commented on docs/design/scheduler.md:
```diff
@@ -0,0 +1,46 @@
+are already running on the node). Second, it applies a set of "priority functions"
+that rank the nodes that weren't filtered out by the predicate check. For example,
+if the PodSpec specified resource limits, then it favors the least-loaded
+machine (measured in terms of the sum of the limits of the Pods already on each node).
```
The occupancy of each node is measured by the sum of the limits of the Pods already on the node divided by the capacity of the node.

davidopp replied:
Good point. I've clarified the wording, also to reflect that we're adding spreading. (Though my intention here was *not* to describe the algorithms in detail -- I think we can have a separate section on that. You are welcome to write that. For this section I just wanted to give a flavor.)

HaiyangDING replied:
The updated description is good enough, simple and clear. I see your points here, and I will try to work out another section for more details like I have said in the comment below.

HaiyangDING commented on docs/design/scheduler.md:
```diff
@@ -0,0 +1,46 @@
+
+## Exploring the code
+
+If you want to get a global picture of how the scheduler work, you can start in
```
"If you want to get a global picture of how the scheduler works, ......"

Hi, I think this is a great design doc for scheduler. I have two other comments above, but I think leaving the doc the way it is now is also a good idea.

davidopp commented:
Addressed the comments, PTAL.

@HaiyangDING Please feel free to follow up with a PR that goes into more detail. I just wanted to make sure we got something simple and high-level in first. (And that would be non-controversial relative to @erictune 's comment :) )

HaiyangDING commented:
@davidopp 

I think it is a great doc, simple and clear. And thank you for sharing your thoughts about this PR:)

I will send a PR to go through the details of the current algorithm and keep updating it once accepted.  The PR will be delivered in a couple of days and by then, it would be really nice if you could give me some advice to improve it.

mikedanese commented:
@davidopp you need to run ./hack/run-gendoc.sh to add the analytics footer.

mikedanese commented:
This is LGTM after you add the analytics footer ^^^

mikedanese commented:
This is LGTM after you add the analytics footer ^^^

mikedanese commented:
@k8s-bot ok to test

davidopp commented:
Thanks, can't believe I forgot that right after I asked about it!

I've run gendocs and squashed the commits. PTAL.

mikedanese commented:
Will add ok-to-merge on shippable green. cc @erictune

davidopp commented:
@k8s-bot ok to test

Dieterbe commented:
@davidopp thanks for your write-up.
for anyone else trying to find the documentation now and not finding it anywhere (the file contributed here was removed, the kubernetes website also doesn't have it for some reason), it lives at:
https://github.com/kubernetes/community/blob/master/contributors/devel/scheduler_algorithm.md
https://github.com/kubernetes/community/blob/master/contributors/devel/scheduler.md
